src/service/src/grpc.rs

```diff
@@ -477,40 +480,85 @@ impl Interceptor for VersionAttachInterceptor {
+#[derive(Clone)]
+struct RequestValidationLayer {
+    version: Version,
+    hostname: Option<String>,
 }
 
-impl VersionCheckExactInterceptor {
-    fn new(version: Version) -> VersionCheckExactInterceptor {
-        VersionCheckExactInterceptor {
-            version: version
-                .to_string()
-                .try_into()
-                .expect("semver versions are valid metadata values"),
+impl<S> tower::Layer<S> for RequestValidationLayer {
+    type Service = RequestValidation<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        let version = self
+            .version
+            .to_string()
+            .try_into()
+            .expect("version is a valid header value");
+        RequestValidation {
+            inner,
+            version,
+            hostname: self.hostname.clone(),
         }
     }
 }
 
-impl Interceptor for VersionCheckExactInterceptor {
-    fn call(&mut self, request: Request<()>) -> Result<Request<()>, Status> {
-        match request.metadata().get(&*VERSION_METADATA_KEY) {
-            None => Err(Status::permission_denied(
-                "request missing version metadata",
-            )),
-            Some(version) if version == self.version => Ok(request),
-            Some(version) => Err(Status::permission_denied(format!(
-                "request version {:?} but {:?} required",
-                version, self.version
-            ))),
+/// A `tower` middleware that validates requests for compatibility with the server.
+#[derive(Clone)]
+struct RequestValidation<S> {
+    inner: S,
+    version: http::HeaderValue,
+    hostname: Option<String>,
+}
+
+impl<S, B> Service<http::Request<B>> for RequestValidation<S>
+where
+    S: Service<http::Request<B>, Error = Box<dyn Error + Send + Sync + 'static>>,
+    S::Response: Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<S::Response, S::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: http::Request<B>) -> Self::Future {
+        let error = |msg| {
+            let error: S::Error = Box::new(Status::permission_denied(msg));
+            Box::pin(future::ready(Err(error)))
+        };
+
+        let Some(req_version) = req.headers().get(VERSION_HEADER_KEY) else {
+            return error("request missing version header".into());
+        };
+        if req_version != self.version {
+            return error(format!(
+                "request has version {req_version:?} but {:?} required",
+                self.version
+            ));
+        }
+
+        let host = req.uri().host();
+        if let (Some(host), Some(name)) = (host, &self.hostname) {
+            if !host.starts_with(name) {
```

Comment:
The `starts_with` here seems weird? Why can't we do a `!=` here?

Reply:
Because `host` is a domain (`foo.bar.com`) and `name` is a hostname (`foo`). We could split `host` at `.` and compare only the first part, but I wasn't sure if there are cases where the hostname actually consists of multiple parts too.

Reply:
Ah! I think we should use FQDNs everywhere. Seems safer. You should be able to get the FQDN hostname inside of Kubernetes by replacing

```
$HOSTNAME
```

with:

```
$(hostname --fqdn)
```

And then it should work to do a straight string equality here.

Reply:
Oh, I like that!